src/ray/core_worker/core_worker.cc

```diff
@@ -326,15 +312,45 @@ Status CoreWorker::Get(const std::vector<ObjectID> &ids, int64_t timeout_ms,
                                                 &result_map, &got_exception));
 
   if (!got_exception) {
+    int64_t local_timeout_ms = timeout_ms;
     if (timeout_ms >= 0) {
-      timeout_ms = std::max(static_cast<int64_t>(0),
-                            timeout_ms - (current_time_ms() - start_time));
+      local_timeout_ms = std::max(static_cast<int64_t>(0),
+                                  timeout_ms - (current_time_ms() - start_time));
     }
-    RAY_RETURN_NOT_OK(memory_store_provider_->Get(memory_object_ids, timeout_ms,
+    RAY_RETURN_NOT_OK(memory_store_provider_->Get(memory_object_ids, local_timeout_ms,
                                                   worker_context_.GetCurrentTaskID(),
                                                   &result_map, &got_exception));
   }
 
+  // If any of the objects have been promoted to plasma, then we retry their
+  // gets at the provider plasma.
+  absl::flat_hash_set<ObjectID> promoted_plasma_ids;
+  for (const auto &pair : result_map) {
+    if (pair.second->IsInPlasmaError()) {
+      promoted_plasma_ids.insert(pair.first.WithTransportType(TaskTransportType::RAYLET));
+    }
+  }
+  if (!promoted_plasma_ids.empty()) {
+    int64_t local_timeout_ms = timeout_ms;
+    if (timeout_ms >= 0) {
+      local_timeout_ms = std::max(static_cast<int64_t>(0),
+                                  timeout_ms - (current_time_ms() - start_time));
+    }
+    RAY_RETURN_NOT_OK(plasma_store_provider_->Get(promoted_plasma_ids, local_timeout_ms,
+                                                  worker_context_.GetCurrentTaskID(),
+                                                  &result_map, &got_exception));
+    for (const auto &id : promoted_plasma_ids) {
+      auto it = result_map.find(id);
+      if (it != result_map.end()) {
+        result_map[id.WithTransportType(TaskTransportType::DIRECT_ACTOR)] = it->second;
+        result_map.erase(id);
```

Comment:
Not sure what you mean. We are getting the data from plasma using the plasma id, and then setting the values for the direct call id, since that's what the client expects as results.

Reply:
Ah I see, you're flipping the flag to plasma to get the IDs, then flipping them back to put them in the map. Can you add that to the comment above `promoted_plasma_ids`?

Reply:
Done